lib/shared-method.js

```diff
@@ -134,100 +151,90 @@ SharedMethod.fromFunction = function(fn, name, sharedClass, isStatic) {
       var desc = accepts[i];
       var name = desc.name || desc.arg;
       var uarg = SharedMethod.convertArg(desc, args[name]);
-      var actualType = SharedMethod.getType(uarg);
-
-      // is the arg optional?
-      // arg was not provided
-      if(actualType === 'undefined') {
-        if(desc.required) {
-          var err = new Error(name + ' is a required arg');
-          err.statusCode = 400;
-          return fn(err);
-        } else {
-          // Add the argument even if it's undefined to stick with the accepts
-          formattedArgs.push(undefined);
-          continue;
-        }
-      }
+      var emptyArray = Array.isArray(uarg) && (uarg.length === 0);
 
-      if(actualType === 'number' && Number.isNaN(uarg)) {
-        var err = new Error(name + ' must be a number');
-        err.statusCode = 400;
-        return fn(err);
+      try {
+        uarg = coerceAccepts(uarg, desc, name);
+      } catch (e) {
+        debug('- %s - ' + e.message, sharedMethod.name);
+        return cb(e);
       }
 
-      // convert strings
-      if(actualType === 'string' && desc.type !== 'any' && actualType !== desc.type) {
-        targetType = desc.type
-        if (Array.isArray(desc.type) && targetType.length == 1) {
-          uarg = uarg.split(/[,|]/)
-          targetType = targetType[0]
-        }
-        switch(targetType) {
-          case 'string':
-            break;
-          case 'date':
-            if (Array.isArray(uarg)) {
-              for (i in uarg) {
-                uarg[i] = new Date(uarg[i]);
-              }
-            } else {
-              uarg = new Date(uarg);
-            }
-            break;
-          case 'number':
-            if (Array.isArray(uarg)) {
-              for (i in uarg) {
-                uarg[i] = Number(uarg[i]);
-              }
-            } else {
-              uarg = Number(uarg);
-            }
-            break;
-          case 'boolean':
-            if (Array.isArray(uarg)) {
-              for (i in uarg) {
-                uarg[i] = Boolean(uarg[i]);
-              }
-            } else {
-              uarg = Boolean(uarg);
-            }
-            break;
-          // Other types such as 'object', 'array',
-          // ModelClass, ['string'], or [ModelClass]
-          default:
-            try {
-              uarg = JSON.parse(uarg);
-            } catch(err) {
-              debug('- %s - invalid value for argument \'%s\' of type \'%s\': %s',
-                sharedMethod.name, name, desc.type, uarg);
-              return fn(err);
-            }
-          break;
+      //Is the arg a valid value?
+      if (desc.required || (SharedMethod.getType(uarg) !== 'undefined' && (!emptyArray))) {
+        var enumValues = desc.enum || (desc.items ? desc.items.enum : null);
+        if (enumValues && doesNotContain(enumValues, uarg)) {
+          var invalidMsg = '';
+          var invalidArgsLen = invalidArgs.length;
+          if (invalidArgsLen < 2) {
```

Comment:
Re-word the error message to not need distinction between singular and plural, then you can just `join` without all this extra logic